### ISSUE TITLE
generate addresses based on xpub

### ISSUE THREAD

**Issue by melaxon** [dan-da/hd-wallet-derive#24]
I took xpub and seed from Electrum and try to reproduce the addresses using hd-wallet-derive.

I run `./hd-wallet-derive.php -g --key=xpub661My......` or
`./hd-wallet-derive.php -g --mnemonic="words from electrum ... blah..blah"`
and for xpub and mnemonic I receive different set of addresses and none of those addresses were found in Electrum

Do I do something wrong?

**Comment by dan-da** [comment]
It looks like you are deriving directly from the key xpub661My....  Try adding --path-preset=electrum.    Or you could instead use --path=m/44'/0'/0'/0/x

If that doesn't work, it could be you are using an older version of electrum.   In which case you could try --path-preset=electrum_legacy.

**Comment by melaxon** [comment]
--path-preset will change nothing
Electrum is the latest version 3.3.8
--path=m/44'/0'/0'/0/x works for mnemonic phrase only but the addresses do not change
  :((

**Comment by melaxon** [comment]
is it possible that index in Electrum starts not from 0 but from e.g. 9999 or something like this?

**Comment by dan-da** [comment]
The flag is actually --preset=<id>.   

That said, I was testing with electrum 3.0.3 right now, and am seeing different addresses.  Also, I tried iancoleman's bip39 tool with the seed from electrum, and it says "invalid mnemonic" which is odd.

KInda busy today.  I will need some time to look into this further.  But will get back to you asap....

**Comment by dan-da** [comment]
Ok, I made a little discovery.   Using electrum 3.0.3:

If I obtain the master private key from electrum's console:

```
>> getmasterprivate()
"xprv9s21ZrQH143K4cAt4WY1KDrHxu48CdvRn4Gjyg7UYmaoWrTDVKfhvTQo3oZJ4abBqpxmF47SH4bGra9GbWuhd5WraxrWYWwstqP2DoehkUA"
```

then hd-wallet-derive generates matching addresses:

```
$ ./hd-wallet-derive.php -g --key=xprv9s21ZrQH143K4cAt4WY1KDrHxu48CdvRn4Gjyg7UYmaoWrTDVKfhvTQo3oZJ4abBqpxmF47SH4bGra9GbWuhd5WraxrWYWwstqP2DoehkUA --cols=path,address --preset=electrum_legacy --numderive=3

+-------+------------------------------------+
| path  | address                            |
+-------+------------------------------------+
| m/0/0 | 156awTpwVBPn8xQULmbCu5eyAwzcJh6GD2 |
| m/0/1 | 1FnU2deqv9pKAPbHY8afc85feoSq8VkQvH |
| m/0/2 | 1MEvdcRjZGDBsmgtpStayFjXM15wNqrfnt |
+-------+------------------------------------+
```

But if I use the mnemonic/seed from electrum, the addresses do not match:

```
$ ./hd-wallet-derive.php -g --mnemonic="merit december trigger right round farm receive level buffalo bounce faculty cruise" --cols=path,address --preset=electrum_legacy --numderive=3

+-------+------------------------------------+
| path  | address                            |
+-------+------------------------------------+
| m/0/0 | 1BARuweKofYP8KUMYm6Am1x1sCoYyDdLrh |
| m/0/1 | 156PEqz6YajFe1awoxrYkZJuGqbRmQWdHC |
| m/0/2 | 1JkQtneJqKMptUpjpjg1wYbndrev7cUDYZ |
+-------+------------------------------------+
```

So I remember now that electrum generates a non bip39 mnemonic.  It can't be used with hd-wallet-derive --mnemonic at present.  see http://docs.electrum.org/en/latest/seedphrase.html

Before closing this, I will add a note to the readme about how to export xprv from electrum because that seems to be the only way.

**Comment by melaxon** [comment]
finally it works with --preset=electrum_legacy and either xprv or xpub
without 'legacy' xprv gives wrong addresses and xpub throws exception:

2019-07-25T22:18:08+03:00 [pid: 83357] [exception] -- #1: Exception. code: 0. Can't derive a hardened key without the private key

**Comment by dan-da** [comment]
Yeah, it seems that legacy is the way to go.  According to a wallet compatibility matrix spreadsheet, linked to from [bitcointalk](https://bitcointalk.org/index.php?topic=1000544.0), electrum 2.6+ switched to bip44.  I just used the paths from that matrix, so haven't tested each wallet software myself.  It seems in this case, the matrix is incorrect, so I will update the code to reflect reality.

The xpub exception is because it is using bip44 paths, which are hardened and therefore cannot be derived from.  I found a comment from the author of electrum stating that's one of the reasons he chose not to use bip44.

thx for bringing this to my attention.

**Comment by foodiesgit** [comment]
for melaxon the syntax you should be using for receive addresses is -g --key=xpub --path=0 and while I'm here dan-da, I can't seem to find a way to give a start position when generating receiveing addresses. for example I use the syntax above with --numderive=10 and it generates the first 10 addresses but I have no idea what parameter I should use to generate the 11th and above while not generating the previous ones too.
